vm/ubpf_vm.c

```diff
@@ -131,14 +133,25 @@ ubpf_exec(const struct ubpf_vm *vm, void *mem, size_t mem_len)
 {
     uint16_t pc = 0;
     const struct ebpf_inst *insts = vm->insts;
-    uint64_t reg[16];
+    uint64_t *reg;
+    bool free_reg;
+    uint64_t retval;
     uint64_t stack[(STACK_SIZE+7)/8];
 
     if (!insts) {
         /* Code must be loaded before we can execute */
         return UINT64_MAX;
     }
 
+    if (vm->regs) {
+        reg = vm->regs;
+        free_reg = false;
+    }
+    else {
+        reg = calloc(NUM_REGS, sizeof(*reg));
```

Comment:
I'm not a big fan of doing a memory allocation each time we run the BPF program. Lots of us are using BPF for performance-critical applications (e.g., userspace networking).

Could you describe your use case for this change?

Reply:
You make a good point. How about if we kept allocation in the stack but allowed to use the external registers?

The use case I have in mind is eBPF offloading. There might be more than one type of memory available and it could be useful to choose where registers should go. It could also help with debugging.

Reply:
> You make a good point. How about if we kept allocation in the stack but allowed to use the external registers?

How do yo propose we do this?

> The use case I have in mind is eBPF offloading. There might be more than one type of memory available and it could be useful to choose where registers should go. It could also help with debugging.

What kind of offloading do you have in mind?

I understand the debugging use case, but I'd rather enable tracing of the uBPF VM than expose new functions to access registers.

Reply:
> How do yo propose we do this?

```
uint64_t _reg[16];
uint64_t *reg;

if (vm->regs)
    reg = vm->regs;
else
    reg = _reg;
```

> What kind of offloading do you have in mind?

Computational storage applications.

> I understand the debugging use case, but I'd rather enable tracing of the uBPF VM than expose new functions to access registers.

What if it was wrapped in a `#ifdef DEBUG`? Then normal operation would not be affected.